dace/transformation/passes/gpustream_scheduling.py

```diff
@@ -0,0 +1,353 @@
+from typing import Union, Dict, Set, List, Tuple
+
+import dace
+from dace import SDFG, properties, SDFGState
+from dace import dtypes
+from dace.codegen import common
+from dace.codegen.targets.experimental_cuda_helpers.gpu_utils import is_within_schedule_types
+from dace.config import Config
+from dace.transformation import pass_pipeline as ppl, transformation
+from dace.sdfg import nodes
+from dace.sdfg.graph import Graph, NodeT
+
+
+@properties.make_properties
+@transformation.explicit_cf_compatible
+class NaiveGPUStreamScheduler(ppl.Pass):
+    """
+    Assigns GPU streams to relevant nodes and inserts synchronization tasklets where needed.
+
+    Strategy Overview:
+    ------------------
+    - GPU stream assignment is based on weakly connected components (WCCs) within each state.
+    - "Relevant nodes" in a WCC are assigned to the same stream.
+        Relevant nodes include:
+          * AccessNodes in GPU memory,
+          * GPU-scheduled nodes (Maps or Library nodes),
+          * Nodes directly connected to the above.
+    - For top-level states (not within nested SDFGs), each new WCC starts on a new stream (starting from 0).
+    - In nested SDFGs:
+        * Stream assignment is inherited from the parent component,
+        * All internal components share the parent's stream (consider revisiting this for performance tuning).
+    - GPU stream IDs wrap around according to the `max_concurrent_streams` configuration.
+    - Synchronization tasklets are inserted using a simple heuristic:
+        * At the end of a state, if outputs certain patterns regarding GPU memory occur,
+        * After a node, if its outputs cross GPU boundaries and are reused downstream.
+
+    Example:
+    --------
+    A state with the following independent chains:
+        K1 → K2
+        K3 → K4 → K5
+        K6
+
+    would be scheduled as:
+        K1, K2     → stream 0
+        K3, K4, K5 → stream 1
+        K6         → stream 2
+
+    (assuming no limit on the number of concurrent streams)
+
+    Note:
+    -----
+    These refer to **backend GPU streams** (e.g., CUDA or HIP), not DaCe symbolic streams.
+    """
+
+    def __init__(self):
+        # max configured number of concurrent streams
+        self._max_concurrent_streams = int(Config.get('compiler', 'cuda', 'max_concurrent_streams'))
+
+        # needed to call correct backend synchronization functions
+        self._backend: str = common.get_gpu_backend()
+
+        # This is expected to be set by the calling backend code generator before applying the pass
+        self._gpu_stream_access_template: str = ""
+
+    def apply_pass(self, sdfg: SDFG, _) -> Dict[nodes.Node, Union[int, str]]:
+        """
+        Assigns GPU streams to nodes and inserts synchronization tasklets where needed.
+        """
+
+        # 1. Traverse each top-level state and assign stream IDs to eligible nodes (starting from stream 0).
```

Comment:
Top-level in the sense of states in the top-level SDFG, I assume? Because `ControlFlowRegion.states()` recurses into sub-ControlFlowRegions, just not into nested SDFGs.